src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -308,10 +315,24 @@ internal EnumerableSorter(Func<TElement, TKey> keySelector, IComparer<TKey> comp
 
         internal override void ComputeKeys(TElement[] elements, int count)
         {
-            _keys = new TKey[count];
-            for (int i = 0; i < count; i++)
+            Func<TElement, TKey> keySelector = _keySelector;
+            if (!ReferenceEquals(keySelector, IdentityFunc))
+            {
+                var keys = new TKey[count];
+                for (int i = 0; i < keys.Length; i++)
+                {
+                    keys[i] = keySelector(elements[i]);
+                }
+                _keys = keys;
+            }
+            else
             {
-                _keys[i] = _keySelector(elements[i]);
+                // The key selector is our known identity function, which means we don't
```

Comment:
Not entirely familiar with the sorting implementation, but does it mutate the contents of `_keys` in any way? Could the fact we're using the `elements` array impact correctness when combining with a secondary key, i.e.
```C#
source.Order().ThenBy(x => ...);
```

Reply:
> but does it mutate the contents of _keys in any way?

No.  You can see in https://github.com/dotnet/runtime/blob/206dccb7945aaa3f26599fbe742de9022ca7ef91/src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs#L299-L324 that _keys is only ever read from after its initialized.  And the poorly-named  internal Sort methods that are handed the `Buffer<>`'s internal buffer don't mutate the array either; they instead produce a sorted map int[] with indices into those elements.

Reply:
SGTM